components/core/src/clp_s/ArchiveReader.cpp

```diff
@@ -92,14 +92,39 @@ SchemaReader& ArchiveReader::read_table(
         throw OperationFailed(ErrorCodeFileNotFound, __FILENAME__, __LINE__);
     }
 
-    auto& schema_reader
-            = create_schema_reader(schema_id, should_extract_timestamp, should_marshal_records);
+    create_schema_reader(
+            m_schema_reader,
+            schema_id,
+            should_extract_timestamp,
+            should_marshal_records
+    );
 
     m_tables_file_reader.try_seek_from_begin(m_id_to_table_metadata[schema_id].offset);
     m_tables_decompressor.open(m_tables_file_reader, cDecompressorFileReadBufferCapacity);
-    schema_reader.load(m_tables_decompressor, m_id_to_table_metadata[schema_id].uncompressed_size);
+    m_schema_reader.load(
+            m_tables_decompressor,
+            m_id_to_table_metadata[schema_id].uncompressed_size
+    );
     m_tables_decompressor.close_for_reuse();
-    return schema_reader;
+    return m_schema_reader;
+}
+
+std::vector<std::shared_ptr<SchemaReader>> ArchiveReader::load_all_tables() {
+    constexpr size_t cDecompressorFileReadBufferCapacity = 64 * 1024;  // 64 KB
+
+    std::vector<std::shared_ptr<SchemaReader>> readers{m_id_to_table_metadata.size(), nullptr};
+    auto rit = readers.begin();
+    for (auto& [id, table_metadata] : m_id_to_table_metadata) {
+        rit->reset(new SchemaReader());
+        create_schema_reader(**rit, id, true, true);
+
+        m_tables_file_reader.try_seek_from_begin(table_metadata.offset);
+        m_tables_decompressor.open(m_tables_file_reader, cDecompressorFileReadBufferCapacity);
+        (*rit)->load(m_tables_decompressor, table_metadata.uncompressed_size);
+        m_tables_decompressor.close_for_reuse();
+        ++rit;
+    }
```

Comment:
```suggestion
    std::vector<std::shared_ptr<SchemaReader>> readers;
    readers.reserve(m_id_to_table_metadata.size());
    for (const auto& [id, table_metadata] : m_id_to_table_metadata) {
        auto schema_reader = std::make_shared<SchemaReader>();
        create_schema_reader(*schema_reader, id, true, true);

        m_tables_file_reader.try_seek_from_begin(table_metadata.offset);
        m_tables_decompressor.open(m_tables_file_reader, cDecompressorFileReadBufferCapacity);
        schema_reader->load(m_tables_decompressor, table_metadata.uncompressed_size);
        m_tables_decompressor.close_for_reuse();

        readers.push_back(std::move(schema_reader));
    }
```
Personally, I would write the code block in this way.. If you want to keep it, do you mind change line 1219 to `*rit = std::make_shared<SchemaReader>();`?

Reply:
I like your way better too.